vowpalwabbit/array_parameters_dense.h

```diff
@@ -114,6 +114,64 @@ class dense_parameters
     for (iterator iter = begin(); iter != end(); ++iter) (&(*iter))[offset] = 0;
   }
 
+  void swap_offsets(size_t offset_1, size_t offset_2, size_t params_per_problem)
+  {
+    float temp = 0.f;
+    assert(offset_1 < offset_2);
+    for (iterator iter = begin(); iter != end(); ++iter)
+    {
+      // assert(*iter == _begin[(count << stride_shift()) & _weight_mask]);
+      size_t actual_index = iter.index() >> stride_shift();
+      size_t actual_type = actual_index & (params_per_problem - 1);
+      if (actual_type == offset_1)
+      {
+        float* other = &_begin[(actual_index + (offset_2 - offset_1)) << stride_shift()];
+
+        if (*other != 0.f || *iter != 0.f)
+        {
+          temp = *other;
+          (&(*other))[0] = (&(*iter))[0];
+          (&(*iter))[0] = temp;
+        }
+      }
+    }
+  }
+
+  void copy_offsets(size_t from, size_t to, size_t params_per_problem)
+  {
+    std::int64_t diff = to - from;
+    // assert(from < to);
+    for (iterator iter = begin(); iter != end(); ++iter)
+    {
+      // assert(*iter == _begin[(count << stride_shift()) & _weight_mask]);
+      size_t actual_index = iter.index() >> stride_shift();
+      size_t actual_type = actual_index & (params_per_problem - 1);
+      if (actual_type == from)
+      {
+        float* other = &_begin[(actual_index + diff) << stride_shift()];
+
+        if (*other != 0.f || *iter != 0.f) { (&(*other))[0] = (&(*iter))[0]; }
+      }
+    }
+  }
+
+  // void copy_offsets(size_t from, size_t to, size_t width = 1, size_t params_per_problem = 1)
+  // {
+  //   for (iterator iter = begin(); iter != end(); ++iter)
+  //   {
+  //     for (size_t j = 0; j < width; j++) { (&(*iter))[to + j] = (&(*iter))[from + j]; }
+  //   }
+  // }
+
+  void clear_offset(size_t offset, size_t params_per_problem)
+  {
+    for (iterator iter = begin(); iter != end(); ++iter)
+    {
+      size_t actual = (iter.index() >> stride_shift()) & (params_per_problem - 1);
+      if (actual == offset && *iter != 0.f) { (&(*iter))[0] = 0; }
```

Comment:
it's going to have to do some extra work with offsets +1,2,3

Reply:
or with 1 << stride_shift() amount of offsets

Reply:
Yeah I've always thought the fact that *iter returns a reference to a single float but we always just turn it back into a pointer and index into it just mega dangerous and icky. Ideally this could be span so the caller of this can't misuse it.

You're using it now thought - so if you're using the rest of the stride then no prob. Feel free to resolve.

Reply:
we can think about it once we have spans, or let me know if you come up with something that fits.